### PR TITLE
[Update] Sample Viewer split view

### PR DESCRIPTION
**Pull request by CalebRas**
## Description

This PR improves the main UI of the Sample Viewer by updating the split view in a few ways:
- Use 3 columns instead of 2.
- Show sidebar by default when the app opens.
- Persist the current category selection.
- Move the search bar/view into an "All" category (more intuitive to use and gets rid of a few bugs).

## Linked Issue(s)

- `swift/issues/5184`

## How To Test

- Test the main UI on all platforms in different rotation modes.

## Screenshots

|Before|After|
|:-:|:-:|
|    Simulator Screenshot - iPad (10th generation) - 2024-04-16 at 16 33 02    |    Simulator Screenshot - iPad (10th generation) - 2024-04-17 at 11 54 18    |

## To Discuss

- I tried to refactor as little as possible, so I added an `AllSamplesView` instead of moving the `searchable` modifier into the `SamplesSearchView`. Let me know if you think I should instead refactor the `SamplesSearchView`.

### CODE REVIEW

File: Shared/Supporting Files/Views/CategoriesView.swift [hunk 1]
```diff
@@ -15,39 +15,74 @@
 import SwiftUI
 
 struct CategoriesView: View {
-    /// The sample categories generated from the samples list.
-    private let categories = Set(SamplesApp.samples.map(\.category)).sorted()
+    /// The visibility of the leading columns in the navigation split view.
+    @Binding var columnVisibility: NavigationSplitViewVisibility
+    
+    /// The selected category loaded from user defaults.
+    @AppStorage("selectedCategory") private var selectedCategory: String?
```
[the remark this answers is not in the document]
Comment (CalebRas) [reply Shared/Supporting Files/Views/CategoriesView.swift:22]: That was the intention, yes. Ting and I decided it wouldn't be a good idea to persist a selected sample in case of potential crashes when loading that would then make the app crash upon opening every time. I figured persisting the category was the next best thing, but I can see where you are coming from. Thoughts @yo1995?

File: Shared/Supporting Files/Views/CategoriesView.swift [hunk 2]
```diff
@@ -15,39 +15,74 @@
...
+                }
+            }
+            .listStyle(.insetGrouped)
+            .navigationTitle(selectedCategory ?? "")
+        }
+        .onChange(of: destinationIsPresented) { _ in
+            // Resets the selection when the navigation destination is no longer presented.
+            guard !destinationIsPresented else { return }
+            selectedCategory = nil
+        }
```
[the remark this answers is not in the document]
Comment (philium) [reply Shared/Supporting Files/Views/CategoriesView.swift:78]: Ah, right, didn't think about that. Never mind.

File: Shared/Supporting Files/Views/CategoriesView.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (yo1995) [reply Shared/Supporting Files/Views/CategoriesView.swift:22]: I'm in favor of not preserving states across multiple app launches. If a user want to have a "shortcut" to a sample, he can favorites it, or don't kill the app and let it stay in the background.

**Review by yo1995 — COMMENTED** [review]
A few comments:

- On iPad landscape, it is not possible to hide the category column, is that correct? I tried the Mails app on iPad and it behaves the same, so I think it is OK.

File: Shared/Supporting Files/Views/CategoriesView.swift [hunk 3]
```diff
@@ -15,39 +15,70 @@
...
+                default:
+                    List(SamplesApp.samples.filter { $0.category == selectedCategory }, id: \.name) {
+                        SampleLink($0)
+                    }
+                }
+            }
+            .listStyle(.insetGrouped)
+            .navigationTitle(selectedCategory ?? "")
+        }
+        .onChange(of: destinationIsPresented) { _ in
```
[the remark this answers is not in the document]
Comment (CalebRas) [reply Shared/Supporting Files/Views/CategoriesView.swift:74]: On Mac Catalyst and iPad yes, is this something you think is needed?

**Comment by CalebRas** [comment]
> On iPad landscape, it is not possible to hide the category column, is that correct? I tried the Mails app on iPad and it behaves the same, so I think it is OK.

Currently, yes, but I think that is intended behavior, which was also present in the Sample Viewer before these changes. If you wanted to hide the category column we could:
1. Use the [prominentDetail](https://developer.apple.com/documentation/swiftui/navigationsplitviewstyle/prominentdetail) navigation split view style, which would overlay the column over the `SampleDetailView` like it does in portrait mode but doing so would prevent using the side-by-side mode and would do the same on Mac Catatlyst.
2. Implement a custom button to hide the category column.

File: Shared/Supporting Files/Views/CategoriesView.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (yo1995) [reply Shared/Supporting Files/Views/CategoriesView.swift:74]: I see. I was just curious if it can be further simplified, but as you pointed out it's needed for iPhone.

**Review by yo1995 — APPROVED** [review]

File: Shared/Supporting Files/Views/CategoriesView.swift [hunk 4]
```diff
@@ -15,39 +15,70 @@
...
+            // Resets the selection when the navigation destination is no longer presented.
+            guard !destinationIsPresented else { return }
+            selectedCategory = nil
+        }
+        .onChange(of: selectedCategory) { newSelection in
+            destinationIsPresented = newSelection != nil
+        }
     }
 }
```
Comment (yo1995) [inline Shared/Supporting Files/Views/CategoriesView.swift:84]: While no change happened to the tile in this PR, I'd suggest setting the overlay's size relative to the background Image, so the icon doesn't take too much space that causes the text to be clipped.

clipped

Since it is not related to the split view changes, ok to do it in a separate PR if you want.

**Review by philium — APPROVED** [review]

File: Shared/Supporting Files/Views/CategoriesView.swift (hunk 4 above)
Comment (CalebRas) [reply Shared/Supporting Files/Views/CategoriesView.swift:84]: Issue: `swift/issues/5250`

**Comment by CalebRas** [comment]
Thank you for your feedback @yo1995 and @philium.
